core/query/sql.ml

```diff
@@ -137,138 +152,162 @@ let order_by_clause n =
-      | UnionAll (qs, n) ->
-        mapstrcat " union all " (fun q -> "(" ^ sq q ^ ")") qs ^ order_by_clause n
-      | Select (fields, [], Constant (Constant.Bool true), _os) ->
-          let fields = string_of_fields fields in
-            "select " ^ fields
-      | Select (fields, [], condition, _os) ->
-          let fields = string_of_fields fields in
-            "select * from (select " ^ fields ^ ") as " ^ fresh_dummy_var () ^ " where " ^ sb condition
-      | Select (fields, tables, condition, os) ->
-          (* using quote_field assumes tables contains table names (not nested queries) *)
-          let tables = List.map (fun (t, x) -> quote t ^ " as " ^ (string_of_table_var x)) tables
-          in string_of_select fields tables condition os
-      | Delete { del_table; del_where } ->
-          string_of_delete del_table del_where
-      | Update { upd_table; upd_fields; upd_where } ->
-          string_of_update upd_table upd_fields upd_where
-      | Insert { ins_table; ins_fields; ins_records } ->
-          string_of_insert ins_table ins_fields ins_records
-      | With (_, q, z, q') ->
-          match q' with
-          | Select (fields, tables, condition, os) ->
-              (* Inline the query *)
-              let tables = List.map (fun (t, x) -> quote t ^ " as " ^ (string_of_table_var x)) tables in
-              let q = "(" ^ sq q ^ ") as " ^ string_of_table_var z in
-              string_of_select fields (q::tables) condition os
-          | _ -> assert false
+  let pr_insert ppf table fields values =
+    let pp_value ppf x = Format.fprintf ppf "(%a)" (Format.pp_print_list ~pp_sep:pp_comma pr_b_ignore_fields) x in
+    Format.fprintf ppf "insert into %a (%a) values %a"
+      pp_constant table
+      (Format.pp_print_list ~pp_sep:pp_comma Format.pp_print_string) fields
+      (Format.pp_print_list ~pp_sep:pp_comma pp_value) values
+  in
+  match q with
+    | UnionAll ([], _) -> Format.fprintf ppf "%a" pp_constant "select 42 as \"@unit@\" where false"
+    | UnionAll ([q], n) -> Format.fprintf ppf "%a%a" pr_q q pp_constant (order_by_clause n)
+    | UnionAll (qs, n) ->
+      let pp_sep_union ppf () = Format.fprintf ppf " union all " in
+      let pp_value ppf x = Format.fprintf ppf "(%a)" pr_q x in
+      Format.fprintf ppf "%a%a"
+        (Format.pp_print_list ~pp_sep:pp_sep_union pp_value) qs
+        pp_constant (order_by_clause n)
+    | Select (fields, [], Constant (Constant.Bool true), _os) ->
+      Format.fprintf ppf "select %a" pr_fields fields
+    | Select (fields, [], condition, _os) ->
+      Format.fprintf ppf "select * from (select %a) as %a where %a"
+        pr_fields fields
+        pp_constant (fresh_dummy_var ())
+        pr_b condition
+    | Select (fields, tables, condition, os) ->
+        (* using quote_field assumes tables contains table names (not nested queries) *)
+        let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
```

Comment:
This is OK for now, but the use of asprintf to serialize an intermediate pretty-printed value to a string and then re-pretty-print it is a symptom of a deeper problem here: the type of `Select` forces tables to be a list of pairs of string table names and variables, but in general selection queries can have subqueries instead of plain table names.  This is not something you have to fix, but I'd appreciate if you could put a FIXME comment here as a reminder to revisit this.

Also, the definition of `tables` is repeated verbatim in the `With` case, so could be hoisted out as a function and reused.